### DISCUSSION TITLE
Object don't move through hole in other object

### DISCUSSION THREAD

**Discussion by DimaRubyAI** [google-deepmind/mujoco#1174] [Asking for Help]
I'm trying to understand what wrong with my scene,
I download a free object from here:
[brake disk](https://www.turbosquid.com/3d-models/brake-disc-3ds-free/1045932)
after small changes that I made in blender I convert it to xml by:
[obj2mjcf](https://github.com/kevinzakka/obj2mjcf) and used this object in [Mujoco-Tutorial](https://github.com/tayalmanan28/MuJoCo-Tutorial).
What I am trying to do is as follows:
Brake disk fall straight in the middle where pole (long and narrow cylinder) located.
I expect that the cylinder go through the disk's hole, but in contrast to that, the disk does not allow penetration through it, as if there is no hole in it.

https://github.com/google-deepmind/mujoco/assets/140307654/08ae329e-a1dc-4b1c-aebd-b63077f3e634

what wrong here?
please let me know if some of the code required to answer this question.

mujoco version: 3.0.0

**Comment by yuvaltassa** [comment]
You need to decompose your mesh into convex pieces. [CoACD](https://github.com/SarahWeiii/CoACD) Is quite good for this.

**Reply by DimaRubyAI** [reply]
Thanks you,
I have tried to use CoACD in the process (before obj2mjcf), with the following parameters:
 coacd -t 0.002 -c 2000
But still getting similar result..

https://github.com/google-deepmind/mujoco/assets/140307654/6c617b0d-8a90-4ed0-a352-ff2ae2e6bb5e

Is there any other reason why this could happen?

**Reply by erez-tom** [reply]
You should visualize the contact meshes which are calculated as the convex hull of the mesh files in the model (the key "h" IIRC).

**Reply by DimaRubyAI** [reply]
Thank you, but I'm not sure that I understand what should I do and what I have tried not really working..
I would appreciate it if you have a link with an example or a tutorial of about it.

**Reply by DimaRubyAI** [reply]
@yuvaltassa, can you please reopen the question or you want me to open a new issue ticket?

**Comment by quagla** [comment]
Another option would be to use the Signed Distance Function plugin https://mujoco.readthedocs.io/en/latest/programming/extension.html#exwriting but it is not available via Python.

**Reply by erez-tom** [reply]
https://github.com/google-deepmind/mujoco/blob/main/plugin/sdf/README.md

**Reply by DimaRubyAI** [reply]
If it is possible at this point I would be happy to stay in Python environment.

**Comment by DimaRubyAI** [comment]
I tried to use official cow example with some small changes (),

- second cow remove
- scale of torus
- partial scale and position of cow1

```
<mujoco>
  <compiler texturedir="asset"/>

  <extension>
    <plugin plugin="mujoco.sdf.torus">
      <instance name="torus">
        <config key="radius1" value="0.15"/>
        <config key="radius2" value="0.05"/>
      </instance>
    </plugin>
    <plugin plugin="mujoco.sdf.sdflib">
      <instance name="sdf">
        <config key="aabb" value="0"/>
      </instance>
    </plugin>
  </extension>

  <asset>
    <texture name="texspot" type="2d" file="spot.png"/>
    <material name="matspot" texture="texspot"/>
    <mesh name="spot" file="asset/spot.obj" scale = "1 2 1">
      <plugin instance="sdf"/>
    </mesh>
    <mesh name="torus" scale = "10 10 10">
      <plugin instance="torus"/>
    </mesh>
  </asset>

  <option sdf_iterations="20" sdf_initpoints="40"/>

  <visual>
    <map force="1000"/>
  </visual>

  <default>
    <geom solref="0.01 1" solimp=".95 .99 .0001" friction="0.5"/>
  </default>

  <statistic meansize="0.2"/>

  <include file="scene.xml"/>

  <worldbody>
    <body pos="0.1 .25 5.7">
      <freejoint/>
      <geom type="sdf" mesh="torus" rgba=".2 .8 .2 1">
        <plugin instance="torus"/>
      </geom>
    </body>
    <body euler="90 0 0" pos="0 0 1.4">
      <geom type="sdf" name="cow1" mesh="spot" material="matspot">
        <plugin instance="sdf"/>
      </geom>
    </body>
    <light name="left" pos="0 0 1"/>
    <light name="right" pos="1 0 1"/>
  </worldbody>
</mujoco>
```

The result is similar to what was seen in the original question


https://github.com/google-deepmind/mujoco/assets/140307654/aeab1746-5fb0-4d6b-be0c-4ca09b068852


What am I doing wrong?
